### PR TITLE
Toggle cluster multinode tests on artery with sys property

### PR DESCRIPTION
**Pull request by johanandren**
By setting system property `akka.cluster.test.use-artery` to `true`
Refs #21204

Local run result:

```
[error] Failed: Total 210, Failed 47, Errors 0, Passed 163
[error] Failed tests:
[error]     akka.cluster.UnreachableNodeJoinsAgain
[error]     akka.cluster.SunnyWeatherArtery
[error]     akka.cluster.RestartFirstSeedNode
[error]     akka.cluster.RestartNode3
[error]     akka.cluster.NodeChurn
[error]     akka.cluster.SplitBrainWithAccrualFailureDetector
[error]     akka.cluster.MemberWeaklyUp
[error]     akka.cluster.QuickRestart
[error]     akka.cluster.SplitBrainWithFailureDetectorPuppet
[error]     akka.cluster.ClientDowningNodeThatIsUpWithAccrualFailureDetector
[error]     akka.cluster.ClusterMetrics
[error]        akka.cluster.Stress
[error]     akka.cluster.SurviveNetworkInstability
```

### CODE REVIEW

File: akka-cluster/build.sbt [hunk 1]
```diff
@@ -9,4 +9,9 @@ Dependencies.cluster
 // disable parallel tests
 parallelExecution in Test := false
 
+// pass on the artery enable/disable flag for the tests
+jvmOptions in MultiJvm ++=
+  (if (sys.props.get("akka.cluster.test.use-artery").exists(_ == "true")) Seq("-Dakka.cluster.test.use-artery=true")
```
Comment (patriknw) [inline akka-cluster/build.sbt]: this is not needed, it's possible to use `-Dakka.remote.artery.enabled=on` directly

Comment (johanandren) [inline akka-cluster/build.sbt]: It passes the props of the sbt JVM to all the multinode JVMs automagically?? O.o

Comment (patriknw) [inline akka-cluster/build.sbt]: yes, all properties starting with `akka` (and some more)

**Comment by johanandren** [comment]
After fixing the shutdown crash race the following tests still fail on my machine (run individually):

* akka.cluster.UnreachableNodeJoinsAgain
* akka.cluster.RestartFirstSeedNode
* akka.cluster.RestartNode3
* akka.cluster.NodeChurn
* akka.cluster.QuickRestart
* akka.cluster.Stress - overloads my machine so that the aeron driver heartbeats times out
* akka.cluster.SurviveNetworkInstability

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiNodeClusterSpec.scala [hunk 2]
```diff
@@ -26,34 +26,43 @@ import akka.actor.RootActorPath
 
 object MultiNodeClusterSpec {
 
+  /** run all tests using artery, rather than the "old" remoting */
+  val useArtery = sys.props.get("akka.cluster.test.use-artery").contains("true")
+  def arteryEnablerFragment = ConfigFactory.parseString("akka.remote.artery.enabled=on")
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiNodeClusterSpec.scala]: This config stuff is still not needed, right?

Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiNodeClusterSpec.scala]: We can point the jenkins job to only run tests in certain projects, e.g. akka-cluster and akka-cluster-tools. Then we should be able to run all with just artery.enabled=on

Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiNodeClusterSpec.scala]: Hmm, yes, I think you are right that just enable it, I expected there to be more stuff that needs to change initially but I haven't found any places where there is yet.

**Comment by johanandren** [comment]
To repeat current stuck-point:
```
$ sbt -Dakka.remote.artery.enabled=on
> akka-cluster/multi-jvm:testOnly akka.cluster.Split
```

**Comment by johanandren** [comment]
Artery-related but PR unrelated failure: https://jenkins.akka.io:8498/job/pr-validator-per-commit-jenkins/7294/consoleFull

**Comment by patriknw** [comment]
I looked at the `SplitSpec`. The difference is that with Artery we get this after healing the partition:
```
[third] [WARN] [09/04/2016 16:40:10.096] [SplitSpec-akka.remote.default-remote-dispatcher-14] [akka.remote.artery.ArteryTransport(akka://SplitSpec)] The remote system artery://SplitSpec@first has quarantined this system artery://SplitSpec@third.
[third] [WARN] [09/04/2016 16:40:10.096] [SplitSpec-akka.remote.default-remote-dispatcher-14] [akka.remote.artery.Association(akka://SplitSpec)] Association to [artery://SplitSpec@first] with UID [-374139991] is irrecoverably failed. Quarantining address. The remote system artery://SplitSpec@first has quarantined this system artery://SplitSpec@third.
```

`first` sends back `Quarantined` when `third` tries to talk to it. In Artery we quarantine it back, which we don't do in old remoting.

See https://github.com/akka/akka/blob/artery-dev/akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala#L589
vs. https://github.com/akka/akka/blob/artery-dev/akka-remote/src/main/scala/akka/remote/Remoting.scala#L489

Quarantining back feels like the right thing to do but perhaps it is risky to do so? Can it result it that the two halves of the cluster continues as two separate clusters?

**Comment by drewhk** [comment]
> which we don't do in old remoting.

Interesting, I thought that we also quarantine in the old remoting. We can safely drop all system message buffers for example at that point as we won't be able to communicate with the other side until it is restarted anyway...

> Quarantining back feels like the right thing to do but perhaps it is risky to do so? Can it result it that the two halves of the cluster continues as two separate clusters?

It is the right way to do, with the only catch, that Quarantine-piercing with Identify messages should work (to be able to ping for a restarted system).

**Comment by patriknw** [comment]
If we want to keep the "quarantine-back" we should have a different assert in the test with Artery. I think that is fine. Rather not touch this behavior in old remoting.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+import scala.util.control.NoStackTrace
+import akka.remote.QuarantinedEvent
+import akka.actor.ExtendedActorSystem
+import akka.remote.RemoteActorRefProvider
+import akka.actor.ActorRef
+import akka.dispatch.sysmsg.Failed
+import akka.actor.PoisonPill
+import akka.actor.Terminated
+
+object SplitMultiJvmSpec extends MultiNodeConfig {
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: does this test add any more coverage than what is already tested in the other SurviveNetwork... ?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+      enterBarrier("downed")
+
+      runOn(side1: _*) {
+        // side2 removed
+        val expected = (side1 map address).toSet
+        awaitAssert {
+          // repeat the downing in case it was not successful, which may
+          // happen if the removal was reverted due to gossip merge, see issue #18767
+          runOn(fourth) {
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: if you keep this test, this is wrong. It should be `runOn(second)`

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+      // side2 should not detect side1 as reachable again
+      Thread.sleep(10000)
+
+      runOn(side1: _*) {
+        val expected = (side1 map address).toSet
+        clusterView.members.map(_.address) should ===(expected)
+      }
+
+      runOn(side2: _*) {
+        if (system.settings.config.getBoolean("akka.remote.artery.enabled")) {
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: `RARP(system).provider.remoteSettings.Artery.Enabled`

File: akka-cluster/src/multi-jvm/scala/akka/cluster/RestartNode3Spec.scala [hunk 6]
```diff
@@ -50,8 +50,12 @@ abstract class RestartNode3Spec
 
   lazy val restartedSecondSystem = ActorSystem(
     system.name,
-    ConfigFactory.parseString("akka.remote.netty.tcp.port=" + secondUniqueAddress.address.port.get).
-      withFallback(system.settings.config))
+    ConfigFactory.parseString(
+      if (system.settings.config.getBoolean("akka.remote.artery.enabled"))
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/RestartNode3Spec.scala]: RARP(system).provider.remoteSettings.Artery.Enabled

File: akka-cluster/src/multi-jvm/scala/akka/cluster/RestartFirstSeedNodeSpec.scala [hunk 7]
```diff
@@ -52,8 +52,12 @@ abstract class RestartFirstSeedNodeSpec
 
   lazy val restartedSeed1System = ActorSystem(
     system.name,
-    ConfigFactory.parseString("akka.remote.netty.tcp.port=" + seedNodes.head.port.get).
-      withFallback(system.settings.config))
+    ConfigFactory.parseString(
+      if (system.settings.config.getBoolean("akka.remote.artery.enabled"))
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/RestartFirstSeedNodeSpec.scala]: RARP(system).provider.remoteSettings.Artery.Enabled

File: akka-cluster/src/multi-jvm/scala/akka/cluster/QuickRestartSpec.scala [hunk 8]
```diff
@@ -72,10 +72,19 @@ abstract class QuickRestartSpec
             else
               ActorSystem(
                 system.name,
-                ConfigFactory.parseString(s"""
-                  akka.cluster.roles = [round-$n]
-                  akka.remote.netty.tcp.port = ${Cluster(restartingSystem).selfAddress.port.get}""") // same port
-                  .withFallback(system.settings.config))
+                // use the same port
+                ConfigFactory.parseString(
+                  if (system.settings.config.getBoolean("akka.remote.artery.enabled"))
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/QuickRestartSpec.scala]: RARP(system).provider.remoteSettings.Artery.Enabled

File: akka-cluster/src/multi-jvm/scala/akka/cluster/UnreachableNodeJoinsAgainSpec.scala [hunk 9]
```diff
@@ -160,18 +160,30 @@ abstract class UnreachableNodeJoinsAgainSpec
 
       runOn(victim) {
         val victimAddress = system.asInstanceOf[ExtendedActorSystem].provider.getDefaultAddress
+        val freshConfig =
+          ConfigFactory.parseString(
+            if (system.settings.config.getBoolean("akka.remote.artery.enabled"))
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/UnreachableNodeJoinsAgainSpec.scala]: RARP(system).provider.remoteSettings.Artery.Enabled

**Comment by patriknw** [comment]
LGTM after some RARP and possibly removing SplitSpec

File: akka-cluster/src/multi-jvm/scala/akka/cluster/QuickRestartSpec.scala (hunk 8 above)
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/QuickRestartSpec.scala]: RARP is `private[remote]` this is `akka.cluster`

Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/QuickRestartSpec.scala]: I see, let's make it `private[akka]` then

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala (hunk 3 above)
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: Hmm. They both do isolate islands, but I can't see that it downs nodes while they are isolated, which would perhaps be a different scenario? Not sure if it is relevant to test though.

Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: but isn't this "identical" to https://github.com/akka/akka/blob/artery-dev/akka-cluster/src/multi-jvm/scala/akka/cluster/SurviveNetworkInstabilitySpec.scala#L302 ?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SurviveNetworkInstabilitySpec.scala [hunk 10]
```diff
@@ -36,6 +36,7 @@ object SurviveNetworkInstabilityMultiJvmSpec extends MultiNodeConfig {
   commonConfig(debugConfig(on = false).withFallback(
     ConfigFactory.parseString("""
       akka.remote.system-message-buffer-size=100
+      akka.remote.artery.advanced.system-message-buffer-size=100
       akka.remote.netty.tcp.connection-timeout = 10s
       """)).
     withFallback(MultiNodeClusterSpec.clusterConfig))
```
Comment (patriknw) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SurviveNetworkInstabilitySpec.scala]: don't you need the special assert in this test also, as you did in SplitSpec?
wasn't the issue in this test originally?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala (hunk 3 above)
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/SplitSpec.scala]: Yes you are right, and it does indeed fail the exact same way, not sure how I could miss that. I'll update that and delete this instead.

**Comment by patriknw** [comment]
@johanandren Have run all tests locally? The cluster tests are not run by the PR validator (not even with old remoting), because I wanted quicker turn-around on the PR validation for this branch.

**Comment by johanandren** [comment]
Yeah, I just realized, I did that until I context-switched this PR out and then back today. Will run locally and ping here once done.

**Comment by patriknw** [comment]
please squash and merge after that, and then we'll setup the separate jenkins job

**Pull request merged**